File: src/main/java/uk/gov/justice/hmpps/prison/aop/connectionproxy/OracleConnectionAspect.java
package uk.gov.justice.hmpps.prison.aop.connectionproxy;

import lombok.extern.slf4j.Slf4j;
import oracle.jdbc.driver.OracleConnection;
import org.apache.commons.lang3.StringUtils;
import org.aspectj.lang.annotation.Aspect;
import org.slf4j.MDC;
import uk.gov.justice.hmpps.prison.security.AuthenticationFacade;
import uk.gov.justice.hmpps.prison.web.config.RoutingDataSource;

import java.sql.Connection;
import java.sql.SQLException;
import java.util.Properties;

import static java.lang.String.format;
import static uk.gov.justice.hmpps.prison.aop.connectionproxy.AppModuleName.MERGE;
import static uk.gov.justice.hmpps.prison.aop.connectionproxy.AppModuleName.PRISON_API;
import static uk.gov.justice.hmpps.prison.security.AuthSource.NOMIS;
import static uk.gov.justice.hmpps.prison.util.MdcUtility.IP_ADDRESS;
import static uk.gov.justice.hmpps.prison.util.MdcUtility.PROXY_USER;
import static uk.gov.justice.hmpps.prison.util.MdcUtility.REQUEST_URI;
import static uk.gov.justice.hmpps.prison.util.MdcUtility.SUPPRESS_XTAG_EVENTS;
import static uk.gov.justice.hmpps.prison.util.MdcUtility.USER_ID_HEADER;

@Aspect
@Slf4j
public class OracleConnectionAspect extends AbstractConnectionAspect {

    private final AuthenticationFacade authenticationFacade;
    private final RoleConfigurer roleConfigurer;
    private final String defaultSchema;

    public OracleConnectionAspect(
        final AuthenticationFacade authenticationFacade,
        final RoleConfigurer roleConfigurer,
        final String defaultSchema) {

        this.authenticationFacade = authenticationFacade;
        this.roleConfigurer = roleConfigurer;
        this.defaultSchema = defaultSchema;
    }

    @Override
    protected Connection openProxySessionIfIdentifiedAuthentication(final Connection pooledConnection) throws SQLException {
        if (!RoutingDataSource.isReplica()) {
            clearContext(pooledConnection);
        }
        if (proxyUserEndpointAndUserSignedIntoNomis()) {
            log.trace("Configuring Oracle Proxy Session for NOMIS user {}", pooledConnection);
            assertNotSlow();
            return openAndConfigureProxySessionForConnection(pooledConnection);
        }
        else if (proxyUserEndpointAndUserIncludedInClientCredentials()) {
            assertNotSlow();
            // If proxy user is set - try to set the context - allow it to fail and carry on
            try {
                setContext(pooledConnection);
            } catch (SQLException e) {
                log.warn("Failed to set context for proxy user {}", MDC.get(PROXY_USER));
            }
        }
        else if ("true".equals(MDC.get(SUPPRESS_XTAG_EVENTS))) {
            assertNotSlow();
            setContextAuditModuleOnly(pooledConnection);
        }

        setDefaultSchema(pooledConnection);
        return pooledConnection;
    }

    private void clearContext(Connection conn) throws SQLException {
        try (final var ps = conn.prepareStatement("BEGIN nomis_context.close_session(); END;")) {
            ps.execute();
        }
    }

    private boolean proxyUserEndpointAndUserSignedIntoNomis() {
        final var proxyUserAuthSource = authenticationFacade.getProxyUserAuthenticationSource();
        return proxyUserAuthSource == NOMIS;
    }

    private boolean proxyUserEndpointAndUserIncludedInClientCredentials() {
        return StringUtils.isNotBlank(MDC.get(PROXY_USER));
    }

    private void assertNotSlow() {
        if (RoutingDataSource.isReplica()) {
            throw new RuntimeException("Found a non read-only transaction annotated with @SlowReportQuery");
        }
    }

    private Connection openAndConfigureProxySessionForConnection(final Connection pooledConnection) throws SQLException {

        final var oracleConnection = openProxySessionForCurrentUsername(pooledConnection);

        final Connection wrappedConnection = new ProxySessionClosingConnection(pooledConnection);

        roleConfigurer.setRoleForConnection(oracleConnection);

        configureConnection(wrappedConnection);

        return wrappedConnection;
    }

    private void configureConnection(final Connection pooledConnection) throws SQLException {
        setDefaultSchema(pooledConnection);
        setContext(pooledConnection);
    }

    private OracleConnection openProxySessionForCurrentUsername(final Connection pooledConnection) throws SQLException {

        final var oracleConnection = (OracleConnection) pooledConnection.unwrap(Connection.class);

        final var info = new Properties();
        final var currentUsername = authenticationFacade.getCurrentUsername();
        info.put(OracleConnection.PROXY_USER_NAME, currentUsername);

        try {
            oracleConnection.openProxySession(OracleConnection.PROXYTYPE_USER_NAME, info);
            log.debug("Opened proxy connection {}", oracleConnection);
        } catch (final SQLException e) {
            log.error("User {} does not support Proxy Connection", currentUsername);
            throw e;
        }
        log.debug("Proxy Connection for {} Successful", currentUsername);
        return oracleConnection;
    }

    private void setDefaultSchema(final Connection conn) throws SQLException {
        if (StringUtils.isNotBlank(defaultSchema)) {
            try (final var ps = conn.prepareStatement(format("ALTER SESSION SET CURRENT_SCHEMA=%s", defaultSchema))) {
                ps.execute();
            }
        }
    }

    private void setContext(final Connection conn) throws SQLException {
        final var sql = format("""
                BEGIN
                %s;
                nomis_context.set_context('AUDIT_USER_ID', '%s');
                nomis_context.set_client_nomis_context('%s', '%s', '%s', '%s');
                END;""",
            callAuditModuleText(),
            MDC.get(USER_ID_HEADER),
            MDC.get(USER_ID_HEADER), MDC.get(IP_ADDRESS), "API", MDC.get(REQUEST_URI));
        try (final var ps = conn.prepareStatement(sql)) {
            log.info("Setting NOMIS context with sql {}", sql);
            ps.execute();
        }
    }

    private void setContextAuditModuleOnly(final Connection conn) throws SQLException {
        final var sql = format("""
                BEGIN
                %s;
                END;""",
            callAuditModuleText());
        try (final var ps = conn.prepareStatement(sql)) {
            log.info("Setting NOMIS context with sql {}", sql);
            ps.execute();
        }
    }

    private String callAuditModuleText() {
        var nomisContext = PRISON_API;
        if ("true".equals(MDC.get(SUPPRESS_XTAG_EVENTS))) {
            nomisContext = MERGE;
        }
        return format("nomis_context.set_context('AUDIT_MODULE_NAME', '%s')", nomisContext);
    }

}

File: src/main/java/uk/gov/justice/hmpps/prison/aop/connectionproxy/OracleConnectionAspect.kt
package uk.gov.justice.hmpps.prison.aop.connectionproxy

import oracle.jdbc.driver.OracleConnection
import org.aspectj.lang.annotation.Aspect
import org.slf4j.LoggerFactory
import org.slf4j.MDC
import org.springframework.context.annotation.Profile
import org.springframework.stereotype.Component
import org.springframework.transaction.TransactionException
import uk.gov.justice.hmpps.kotlin.auth.AuthSource
import uk.gov.justice.hmpps.kotlin.auth.AuthSource.NOMIS
import uk.gov.justice.hmpps.kotlin.auth.HmppsAuthenticationHolder
import uk.gov.justice.hmpps.prison.util.MdcUtility.IP_ADDRESS
import uk.gov.justice.hmpps.prison.util.MdcUtility.PROXY_USER
import uk.gov.justice.hmpps.prison.util.MdcUtility.REQUEST_URI
import uk.gov.justice.hmpps.prison.util.MdcUtility.SUPPRESS_XTAG_EVENTS
import uk.gov.justice.hmpps.prison.util.MdcUtility.USER_ID_HEADER
import uk.gov.justice.hmpps.prison.web.config.RoutingDataSource
import java.lang.Exception
import java.sql.Connection
import java.sql.SQLException
import java.util.Properties

@Aspect
@Component
@Profile("connection-proxy")
class OracleConnectionAspect(
  private val hmppsAuthenticationHolder: HmppsAuthenticationHolder,
  private val roleConfigurer: RoleConfigurer,
  private val nomisConfigurer: NomisConfigurer,
) : AbstractConnectionAspect() {

  private val log = LoggerFactory.getLogger(this::class.java)

  @Throws(SQLException::class, ProxyConnectionTransactionException::class)
  public override fun configureNomisConnection(pooledConnection: Connection): Connection = with(pooledConnection) {
    when {
      isNomisProxyUser() -> openProxySessionConnection()
      isProxyUser() -> openResettableContextConnection()
      isSuppressXTags() -> openXtagsSuppressingConnection()
      else -> this.also { nomisConfigurer.setDefaultSchema(it) }
    }
  }

  @Throws(SQLException::class)
  private fun Connection.openProxySessionConnection(): Connection {
    log.debug("Configuring Oracle Proxy Session")
    assertNotSlow()
    openProxySessionForCurrentUsername()
      .also { oracleConnection -> roleConfigurer.setRoleForConnection(oracleConnection) }
    return ProxySessionClosingConnection(this)
      .also { proxySessionConnection ->
        nomisConfigurer.setDefaultSchema(proxySessionConnection)
        nomisConfigurer.setNomisContext(proxySessionConnection, mdc(USER_ID_HEADER), mdc(IP_ADDRESS), mdc(REQUEST_URI), "APP", isSuppressXTags())
      }
  }

  @Throws(SQLException::class)
  private fun Connection.openProxySessionForCurrentUsername(): OracleConnection {
    val currentUsername = hmppsAuthenticationHolder.principal
    val info = Properties().apply {
      this[OracleConnection.PROXY_USER_NAME] = currentUsername
    }

    return (this.unwrap(Connection::class.java) as OracleConnection)
      .apply {
        try {
          openProxySession(OracleConnection.PROXYTYPE_USER_NAME, info)
        } catch (e: SQLException) {
          if (e.errorCode == 1017) throw ProxyConnectionTransactionException("Could not open proxy session as $currentUsername - account doesn't support proxy connections", e)
          if (e.errorCode == 28000) throw ProxyConnectionTransactionException("Could not open proxy session as $currentUsername - account is locked", e)
          throw e
        }
      }
  }

  @Throws(SQLException::class)
  private fun Connection.openResettableContextConnection(): Connection {
    log.info("Configuring session for client credentials user")
    assertNotSlow()
    return ResettableContextConnection(this)
      .also { connection ->
        nomisConfigurer.setDefaultSchema(connection)
        nomisConfigurer.setNomisContext(connection, mdc(USER_ID_HEADER), mdc(IP_ADDRESS), mdc(REQUEST_URI), "APP", isSuppressXTags())
      }
  }

  @Throws(SQLException::class)
  private fun Connection.openXtagsSuppressingConnection(): Connection {
    log.debug("Configuring session to suppress XTag events")
    assertNotSlow()
    return ResettableContextConnection(this)
      .also { connection ->
        nomisConfigurer.setDefaultSchema(connection)
        nomisConfigurer.setSuppressXtagEvents(connection)
      }
  }

  private fun assertNotSlow() {
    if (RoutingDataSource.isReplica()) {
      throw RuntimeException("Found a non read-only transaction annotated with @SlowReportQuery")
    }
  }

  private fun isNomisProxyUser(): Boolean = isProxyUser() && authSource() == NOMIS

  private fun isProxyUser(): Boolean = !mdc(PROXY_USER).isNullOrBlank()

  private fun isSuppressXTags(): Boolean = "true" == MDC.get(SUPPRESS_XTAG_EVENTS)

  private fun authSource(): AuthSource = hmppsAuthenticationHolder.authSource

  private fun mdc(key: String): String? = MDC.get(key)
}

@Throws(SQLException::class)
internal fun Connection.run(sql: String) = prepareStatement(sql).use { ps -> ps.execute() }

class ProxyConnectionTransactionException(message: String, e: Exception) : TransactionException(message, e)
